src/character.h

```diff
@@ -2929,9 +2922,11 @@ class Character : public Creature, public visitable
 
         struct weighted_int_list<std::string> melee_miss_reasons;
 
-        int cached_moves;
-        tripoint cached_position;
-        pimpl<inventory> cached_crafting_inventory;
+        /* crafting inventory cached time */
+        mutable time_point cached_time;
+        mutable int cached_moves;
+        mutable tripoint cached_position;
+        mutable pimpl<inventory> cached_crafting_inventory;
```

Comment:
If these are all related to the cached crafting inventory then it might be cleaner to separate them out into a new `struct` and then you could have just one mutable member of `Character` which is that new `struct`.  OTOH, that's probably a lot of churn which maybe should be a separate PR.

Reply:
Done. I considered doing that earlier but wasn't sure where to put the struct definition. I've now added it in-place since I've seen that being done in other places. Also noticed that the "radius" cache parameter was stored in a static var instead of a field for reasons I could not comprehend so I added it to the struct.
Do you think it's worth adding any methods (like reset/invalidate) to the struct?

Reply:
> Do you think it's worth adding any methods (like reset/invalidate) to the struct?

Potentially.  If such code is currently duplicated, or there are `Character` methods that only refer to these fields, then they could sensibly be made a method of the struct.  Go with whatever you think makes the code most readable.

Reply:
Ok, I'll leave it as is. The only "duplication" is between Character::Character() and Character::invalidate_crafting_inventory(), both of which call `crafting_cache.time = calendar::before_time_starts;` which seems reasonable to me.